addons/sale_product_configurator/controllers/main.py

```diff
@@ -175,6 +175,10 @@ def sale_product_configurator_get_optional_products(
         parent_combination = request.env['product.template.attribute.value'].browse(
             parent_combination + combination
         )
+        # avoid recursive optional products
+        optional_products = product_template.optional_product_ids.filtered(
+            lambda p: product_template.id not in p.optional_product_ids.ids
+        )
```

Comment:
You also need to make this change in `get_product_configurator_values`

Reply:
No, because it would not show the optional product at all.

Reply:
To maintain a consistent behavior as we discussed I added filtering in get_product_configurator_values as well.